dynamiq/storages/vector/pgvector/pgvector.py

```diff
@@ -0,0 +1,648 @@
+            client (Optional[PGVector]): PGVector client instance. Defaults to None.
+            create_extension (bool): Whether to create the vector extension (if it does not exist). Defaults to True.
+            table_name (str): Name of the table in the database. Defaults to None.
+            schema_name (str): Name of the schema in the database. Defaults to None.
+            dimension (int): Dimension of the embeddings. Defaults to 1536.
+            vector_function (PGVectorVectorFunction): The vector function to use for similarity calculations.
+                Defaults to 'cosine_similarity'.
+            index_method (PGVectorIndexMethod): The index method to use for the vector store.
+                Defaults to 'exact_nearest_neighbor_search'.
+            index_name (str): Name of the index to create. Defaults to None.
+            create_if_not_exist (bool): Whether to create the table and index if they do not exist. Defaults to True.
+            content_key (Optional[str]): The field used to store content in the storage. Defaults to 'content'.
+            embedding_key (Optional[str]): The field used to store embeddings in the storage. Defaults to 'embedding'.
+        """
+        if vector_function not in PGVectorVectorFunction:
+            raise ValueError(f"vector_function must be one of {PGVectorVectorFunction.list()}")
+        if index_method is not None and index_method not in PGVectorIndexMethod:
+            raise ValueError(f"index_method must be one of {PGVectorIndexMethod.list()}")
+
+        if isinstance(connection, str):
+            self.connection_string = connection
+            self._conn = psycopg.connect(self.connection_string)
+        elif isinstance(connection, PGVector):
+            self._conn = connection.connect()
+            self.connection_string = connection.conn_params
+        else:
+            raise ValueError("connection must be a string or PGVector object")
+
+        self.create_extension = create_extension
+        if self.create_extension:
+            self._conn.execute("CREATE EXTENSION IF NOT EXISTS vector")
+            self._conn.commit()
+
+        register_vector(self._conn)
+
+        self.client = client if client else self._conn
+
+        self.table_name = table_name if table_name else DEFAULT_TABLE_NAME
+        self.schema_name = schema_name if schema_name else DEFAULT_SCHEMA_NAME
+        self.dimension = dimension
+        self.index_method = index_method
+        self.vector_function = vector_function
+
+        self.content_key = content_key
+        self.embedding_key = embedding_key
+
+        if (
+            self.index_method == PGVectorIndexMethod.IVFFLAT
+            and self.vector_function == PGVectorVectorFunction.L1_DISTANCE
+        ):
+            msg = "IVFFLAT index does not support L1 distance metric"
+            raise VectorStoreException(msg)
+
+        if create_if_not_exist:
+            with self._get_connection() as conn:
```

Comment:
Why do we again try to connect and don't reuse self.client or self._conn?

Reply:
`_get_connection()` will just yield `self._conn` if it is already established and only tries to connect when the connection is closed or absent (as far as I understand it).